### PR TITLE
feat: backend changes for updating run behaviour to automatic

### LINKED ISSUE #40468: [Task]: Update backend logic to send runBehaviour in actionUpdates when binding / unbinding happens
### SubTasks

When backend automatically updates the value of runBehaviour based on binding / unbinding in the update page layout API, we also need to have a way to propagate these updates to frontend so that frontend can update the state in the UI accordingly. This is done using `actionUpdates` property in the page layout update response.

With current implementation, it sends executeOnLoad in `actionUpdates` (runBehaviour with milestone 1), In this task, we need to send this value appropriately based on flags

- If flag is true, send runBehaviour to be automatic in actionUpdates on binding. Send runBehaviour to be manual in actionUpdates on removal of binding
- If flag is false, send runBehaviour to be page load in actionUpdates on binding. Send runBehaviour to be manual in actionUpdates on removal of binding

### PR DESCRIPTION
**Pull request by sneha122**
## Description
This PR adds backend changes for making run behaviour either of Automatic (Which means query will be triggered every time its dependency changes) or Page load (Query executed only on page load) based on feature flag.

Fixes #40468 #40467 
_or_  
Fixes `Issue URL`
> [!WARNING]  
> _If no issue exists, please create an issue first, and check with the maintainers if the issue is valid._

## Automation

/ok-to-test tags="@tag.Datasource"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/15112643740>
> Commit: 8fe5458cbf5f5951c22083ff15ff1ed019188246
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=15112643740&attempt=1" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.Datasource`
> Spec:
> <hr>Mon, 19 May 2025 12:55:41 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [ ] No

## Summary by CodeRabbit

- **New Features**
  - Introduced a feature flag to enable reactive actions.
  - Added an automatic run behavior for actions to trigger on dependency changes when enabled.

- **Enhancements**
  - Enhanced action execution configurability based on feature flag status.

- **Tests**
  - Added tests verifying action run behavior updates respecting the new feature flag and user settings.

### CODE REVIEW

File: app/server/appsmith-server/src/main/java/com/appsmith/server/onload/internal/OnLoadExecutablesUtilCEImpl.java [hunk 1]
```diff
@@ -320,103 +323,116 @@ public Mono<Boolean> updateExecutablesRunBehaviour(
...
-                        if (FALSE.equals(executable.getUserSetOnLoad())) {
-
-                            // If this executable is no longer an onload executable, turn the execute on load to false
-                            if (turnedOffExecutableNames.contains(executableName)) {
-                                executable.setRunBehaviour(RunBehaviourEnum.MANUAL);
-                                toUpdateExecutables.add(executable);
-                            }
+                            Set<String> newOnLoadExecutableNames = onLoadExecutables.stream()
+                                    .map(Executable::getUserExecutableName)
+                                    .collect(Collectors.toSet());
```
[the remark this answers is not in the document]
Comment (sneha122) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/onload/internal/OnLoadExecutablesUtilCEImpl.java:353]: Thanks for the callout @subrata71 I will create EE PR

File: app/server/appsmith-server/src/main/java/com/appsmith/server/onload/internal/OnLoadExecutablesUtilCEImpl.java [hunk 2]
```diff
@@ -320,103 +323,116 @@ public Mono<Boolean> updateExecutablesRunBehaviour(
...
-                            // should not be updated
-                            turnedOnExecutableNames.remove(executableName);
-                            turnedOffExecutableNames.remove(executableName);
-                        }
-                    }
+                                    // If this executable is newly found to be on load, turn execute on load to true
+                                    if (turnedOnExecutableNames.contains(executableName)) {
+                                        // Use the prefetched feature flag value
+                                        if (isReactiveActionsEnabled) {
+                                            executable.setRunBehaviour(RunBehaviourEnum.AUTOMATIC);
```
[the remark this answers is not in the document]
Comment (sneha122) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/onload/internal/OnLoadExecutablesUtilCEImpl.java:385]: I am planning to add unit test cases for this change now in this PR but we also have a plan to embed e2e test cases once complete feature is developed. You can track that task [here](https://github.com/appsmithorg/appsmith/issues/40656)

File: app/server/appsmith-server/src/main/java/com/appsmith/server/onload/internal/OnLoadExecutablesUtilCEImpl.java [hunk 3]
```diff
@@ -320,103 +323,116 @@ public Mono<Boolean> updateExecutablesRunBehaviour(
...
-                            // Remove the executable name from either of the lists (if present) because this executable
-                            // should not be updated
-                            turnedOnExecutableNames.remove(executableName);
-                            turnedOffExecutableNames.remove(executableName);
-                        }
-                    }
+                                    // If this executable is newly found to be on load, turn execute on load to true
+                                    if (turnedOnExecutableNames.contains(executableName)) {
+                                        // Use the prefetched feature flag value
+                                        if (isReactiveActionsEnabled) {
```
[the remark this answers is not in the document]
Comment (sneha122) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/onload/internal/OnLoadExecutablesUtilCEImpl.java:384]: We won't need auto commit changes for this right? We are not introducing new prop now, we are introducing new state for that new prop we added in milestone 1

File: app/server/appsmith-server/src/main/java/com/appsmith/server/onload/internal/OnLoadExecutablesUtilCEImpl.java [hunk 4]
```diff
@@ -320,103 +323,116 @@ public Mono<Boolean> updateExecutablesRunBehaviour(
...
 
-                    // There are no actions in this page. No need to proceed further since no actions would get updated
-                    if (creatorContextExecutables.isEmpty()) {
-                        return Mono.just(FALSE);
-                    }
+                            // No actions require an update if no actions have been found as page load actions as well
+                            // as
+                            // existing on load page actions are empty
+                            if (existingOnLoadExecutables.isEmpty()
+                                    && (onLoadExecutables == null || onLoadExecutables.isEmpty())) {
```
[the remark this answers is not in the document]
Comment (sneha122) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/onload/internal/OnLoadExecutablesUtilCEImpl.java:342]: Thanks Subrata, will make the changes

File: app/server/appsmith-server/src/main/java/com/appsmith/server/onload/internal/OnLoadExecutablesUtilCEImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (subrata71) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/onload/internal/OnLoadExecutablesUtilCEImpl.java:384]: There is a case which can cause diff on top of the clean state of the app.

Let’s assume the app is connected to Git and an action is set to run on page load (`ON_PAGE_LOAD` in the DB) with no diffs. When the feature flag for `AUTOMATIC` run behavior is enabled, and the user makes changes that turn off the on-page load setting, the DB updates to `MANUAL`. After discarding changes and pulling the latest commit, the app should have no diffs. However, the feature flag causes the run behavior to switch to `AUTOMATIC` during reconstruction, resulting in an unexpected diff. This is the specific edge case I’m referring to.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/onload/internal/OnLoadExecutablesUtilCEImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (sneha122) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/onload/internal/OnLoadExecutablesUtilCEImpl.java:355]: EE PR: https://github.com/appsmithorg/appsmith-ee/pull/7546

File: app/server/appsmith-server/src/main/java/com/appsmith/server/onload/internal/OnLoadExecutablesUtilCEImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sneha122) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/onload/internal/OnLoadExecutablesUtilCEImpl.java:386]: That's a very valid concern and thanks for raising this!! @subrata71 

@infinitetrooper @pedro-santos-rodrigues Can you please check this edge case and let me know what should happen in this case?

I have triggered EE DP [here](https://github.com/appsmithorg/appsmith-ee/pull/7546) in case you need to test this scenario

CC: @vivek-appsmith @ankitakinger

File: app/server/appsmith-server/src/main/java/com/appsmith/server/onload/internal/OnLoadExecutablesUtilCEImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (sneha122) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/onload/internal/OnLoadExecutablesUtilCEImpl.java:387]: Unit tests have been added

File: app/server/appsmith-server/src/main/java/com/appsmith/server/onload/internal/OnLoadExecutablesUtilCEImpl.java [hunk 5]
```diff
@@ -331,15 +331,15 @@ public Mono<Boolean> updateExecutablesRunBehaviour(
 
                             // There are no actions in this page. No need to proceed further since no actions would get
                             // updated
-                            if (creatorContextExecutables.isEmpty()) {
+                            if (creatorContextExecutables.isNullOrEmpty()) {
```
Comment (subrata71) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/onload/internal/OnLoadExecutablesUtilCEImpl.java]: Can you please use `com.appsmith.server.helpers.CollectionUtils.isNullOrEmpty()` instead?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/onload/internal/OnLoadExecutablesUtilCEImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (subrata71) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/onload/internal/OnLoadExecutablesUtilCEImpl.java:386]: @pedro-santos-rodrigues Could you please review the scenario I shared? Do you agree that, in this specific case, it would lead to uncommitted changes? If so, do you think this is something we should be concerned about?

**Review by subrata71 — APPROVED** [review]
[This discussion](https://github.com/appsmithorg/appsmith/pull/40671#discussion_r2095498608) can go in parallel. Approving this as it's not an immediate blocker for this PR.

**Comment by sneha122** [comment]
Not merging this PR because its equivalent PR has few [cypress tests failing](https://github.com/appsmithorg/appsmith-ee/pull/7546)

**Comment by sneha122** [comment]
All tests have now passed on equivalent [EE PR](https://github.com/appsmithorg/appsmith-ee/pull/7546) as well
